### DISCUSSION TITLE
Can we deactivate /page/ functionality from blog functionality?

### DISCUSSION THREAD

**Discussion by johnnyreilly** [facebook/docusaurus#8311] [General]
The "Older Entries" / "Newer Entries" functionality at the bottom of blog posts leads to the creation of "pseudo" pages which allow you to step through your history of pages and duplicate blog content along the way.

Historically this was useful for SEO as it meant there was a simple way to crawl all content. 

Given that we have a blog archive page in Docusaurus now, I wonder if it might be possible to turn off the generation of this.  All the links that are required are covered by the sitemap and the blog archive, so is there a need for these pseudo pages?

Could we have a config option to turn them off perhaps?

Or do they cover a use case I've not considered?
<img width="701" alt="2022-11-09 15_00_28-Page indexing" src="https://user-images.githubusercontent.com/1010525/200870991-26269d02-782d-42dd-b647-294a871d79f9.png">

Raised this on Canny also in case it belongs there: https://docusaurus.io/feature-requests/p/can-we-deactivate-page-functionality-from-blog-functionality

**Comment by slorber** [comment]
The problem is that most sites don't link to the archive page, and yet they want their older posts to remain available without having an unlimited sidebar nor a "blog archive" navbar item.

We should find a good story for those sites.

Maybe having a `blogHome` feature with the ability to select a list of featured blog posts could make sense? Many would prefer featuring best blog posts rather than the X last ones.

**Reply by johnnyreilly** [reply]
Oh, I'm not suggesting getting rid of pagination entirely, more just having the option to turn it off

But is there a reason we couldn't turn off the "older entries / newer entries" (and consequently "/pages/" ) functionality for those it's not useful to?  (me 😄 - but maybe not only me)

**Reply by slorber** [reply]
No particular reason apart it's a bit confusing to only show the first page without the ability to navigate to the subsequent ones.

To me what you want is a clean, better looking blog home, so thinking from first principles, maybe it's better to fully disable the pagination (including first page) and give you a flexible API to implement a good home (that can display the 5 first blog posts if you want, but you can enhance it later)

**Reply by johnnyreilly** [reply]
It's probably worth me talking about how I navigate my blog.

I navigate my blog in the following ways:
1. I go to the blog and look at the latest entry (often what I'm interested in) **or**
2. I click on the search box and type words in to get the algoiia search results that leads me to the post I'm after **or**
3. I click on "archive" and hit CMD+F to find a post I'm interested in **or**
4. I google my name and the topic I'm after and pick a blog post out of the search results 

The thing I never do is use the pagination mechanism. I often can't remember when I wrote something, and I write a lot. I'd spend ages clicking "older entries" and scrolling. So the pagination mechanism isn't useful to me.

In the end I actually really like the blogs behaviour as it is. I don't really want it to be different apart from one thing: can we turn pagination off?

With it off:
- I would have less pages on the blog, and since we'd be removing duplicate /pages/ style pages so that's good. Website size would decrease, data costs would decrease, search engines would have less to crawl.
- I'd get rid of functionality I don't use (or consider very useful)
- possibly build times might be faster, as less pages are being generated?

Without having looked at the code, it feels like perhaps this might not be too complicated; just turning off certain code paths based on a parameter.

Does what I'm saying make sense?

**Reply by johnnyreilly** [reply]
Proposed implementation here: https://github.com/facebook/docusaurus/pull/8326 (pretty small change as it turned out)

There's a crucial detail I missed out; I run my blog with a single blog post per page: https://blog.johnnyreilly.com/

**Reply by slorber** [reply]
Hey

I understand your use-case, but what I understand is that it's more a technical optimisation than a real feature for users. Some of your visitors might like this pagination, and it's probably not good to decide for them 🤷‍♂️

> There's a crucial detail I missed out; I run my blog with a single blog post per page

That's a quite specific way to use the Docusaurus blog. I can understand how it makes sense to do that, and at the same time I'm not sure it's ideal because the exact same content will be at 2 different addresses with 2 different canonical URLs and Google/SEO might not like that much. 

You'd rather have nothing at `/` instead, or make it target the real blog post canonical URL. 

Same for all the next pages (that you want to deactivate).


----


My main concerns:
- `blogPaginationActive: false` for me doesn't mean keeping only 1 page. It means that you don't want to paginate and more likely render ALL your blog posts on the landing page (except for you, because you want something very specific)
- `postsPerPage` is not an API designed to control how much content should go in blog home landing page

I mean, I understand how this can enable you to achieve what you want, but at the same time it's bad API design and it would apply only for users that have exactly your very specific use-case, and be confusing for others. 

And the paginated list component was not either designed to be used the way you use it (cf the duplicate content / canonical URL problem)

If I were to introduce `blogPaginationActive: false`, I would simply disable ALL pages, not just the 2-n paginated pages. Isn't it confusing for swizzle users if the paginated theme view remains used despite the pagination being turned off? I think it is.

So, if we introduce `blogPaginationActive: false`, now the question is how do we enable users to create a proper blog home, including your use-case of duplicated blog post (which should have the proper canonical URL) but also more advanced cases where you can display your best writings:

<img width="919" alt="CleanShot 2022-11-17 at 15 28 43@2x" src="https://user-images.githubusercontent.com/749374/202472899-69088c92-07e0-4ab9-a8e9-44f8fbd18e44.png">

**Reply by slorber** [reply]
My suggestion would be:


`blogPaginationActive: false`: disable the full blog pagination, including the 1st page


And you can use this on your home blog post:

```md
---
slug: /2022/11/17/azure-ad-claims-static-web-apps-azure-functions
aliases: ["/"] # NEW FEATURE
---

# Azure AD Claims with Static Web Apps and Azure Functions

Text
```

The aliasing feature would create extra pages for another slug (that can be /)

It gives the ability to keep a shared canonical URL for both pages and avoid SEO content being flagged as duplicate.

It also enables you to define which blog post exactly you want to be your home, instead of being forced to display the very last one. Users that want to display their best blog post on their home instead of their latest should be able to do so.

If you want to automatically publish a specific blog post on your home, that can probably be automated later thanks to an upcoming [createFrontMatter](https://github.com/facebook/docusaurus/issues/5568), enabling you to create your own automated rules to define what should be the home, including custom algos more complex than "the last blog post"

Does it make sense?

**Reply by johnnyreilly** [reply]
I think much of what you're suggesting makes sense. I'd like to flesh out the details a little.  

- `blogPaginationActive: false` - disable the full blog pagination, including the 1st page.  I think from what @Josh-Cena [says here](https://github.com/facebook/docusaurus/pull/8326#issuecomment-1318730542) he'd rather that the API was `postsPerPage: 0` to achieve the same end.  This feels less explicit to me than `blogPaginationActive: false`, but I don't feel super strongly about it.  Maybe 2 out of 5. I think `postsPerPage: 0` would be a reasonable API - would be happy to go with it.
- `aliases` - a new feature that allows you to specify the root page.  I think this is really cool - I like the control that this gives; it allows you to have other pages as your "home" page. Selfishly I'd quite like the possibility of an opt in feature that supports having a default alias of the latest post.  To save manually moving `aliases: ["/"]` from the second latest post, to the latest post.  I think you're suggesting that [`createFrontMatter`](https://github.com/facebook/docusaurus/issues/5568) might solve this in future?

I'd be happy to have a go at this implementation if this sounds reasonable?

One more thing to chuck into the mix is the `tags` functionality.  I also don't get much value out of tags - and they're similarly implemented.  How would you feel about a `tagsPaginationActive: false` or something which if set deactivated those?  Happy to decouple that from this conversation, but I thought it might be worth mentioning in case that inspires further thoughts?

**Reply by slorber** [reply]
> I think from what @Josh-Cena https://github.com/facebook/docusaurus/pull/8326#issuecomment-1318730542 he'd rather that the API was postsPerPage: 0 to achieve the same end. This feels less explicit to me than blogPaginationActive: false

`postsPerPage: 0`is less explicit, but at the same time it prevents weird impossible combinations like `postsPerPage: 10` + `blogPaginationActive: false`, so I also don't know what is best, fine with both 👍 

---

> Selfishly I'd quite like the possibility of an opt in feature that supports having a default alias of the latest post.

I understand. Wonder I wonder how we can make it happen.

Later we'll want more flexible options to create a custom homepage, so maybe this API makes sense?

```js
blogHome: {type: "latestPost"}

blogHome: {type: "post",slug: "/2022/11/17/azure-ad-claims-static-web-apps-azure-functions"}

blogHome: "@site/src/MyCustomBlogHome"
```

---

> I'd be happy to have a go at this implementation if this sounds reasonable?

Yes, probably worth splitting into multiple PRs

---

> One more thing to chuck into the mix is the tags functionality. I also don't get much value out of tags - and they're similarly implemented. How would you feel about a tagsPaginationActive: false or something which if set deactivated those? Happy to decouple that from this conversation, but I thought it might be worth mentioning in case that inspires further thoughts?

Good point, forgot about the tags 😅 


https://blog.johnnyreilly.com/tags

It's not clear to me what's the value for you to have tags in the first place, if there's no convenient way to browse the full list of tags, and then see all posts having that specific tag. 


I mean, 1 post per page + full blog post without any truncation marker makes this unusable:
https://blog.johnnyreilly.com/tags/net

Remember users can click on tags to access that page. Do we really want to make blog post tags not clickable and disable all the rest? Then what's the value of having those tags in the first place?

Doesn't it make sense that you have both:
- full latest blog post on blog at /
- paginated lists for tags + truncation markers

In this case we would probably need to have a separate `tagsPostsPerPage` option

**Reply by johnnyreilly** [reply]
So I don't use tag pages at all - they're just there.  I didn't actually realise they generated pages when I first started using Docusaurus! I thought they just created meta tags and badges on a post.  I'd rather they weren't clickable things with navigation associated.  I'd like to lose all "tags pages" and just have them being meta tags on a page perhaps.  I think tag functionality generally makes sense; but I just don't ever use it.  I feel a lot less strongly about it than I do about blog pagination though 😄 

> Later we'll want more flexible options to create a custom homepage, so maybe this API makes sense?
>
> `blogHome: {type: "latestPost"}`

This feels like it has potential, yes.

Happy to roll with `postsPerPage: 0` and make my first PR the deactivating of blog pagination pages.  Then do `aliases` in a subsequent PR.  I don't think it is a usable feature until both are done, but happy to split to avoid mega PRs.

**Reply by johnnyreilly** [reply]
Actually I might do `aliases` first as it offers value to all immediately.

**Reply by johnnyreilly** [reply]
I've put up a first draft of `aliases` here: https://github.com/facebook/docusaurus/pull/8356 - however I'll need a little advice to take it further.

As an aside I did a little experiment last night: https://github.com/johnnyreilly/blog.johnnyreilly.com/pull/344

I built a post-build mechanism that processes `sitemap.xml` and filters out URLs that contain `/page/` and `/tag/` on the basis that these mostly represent duplicate content to Google.  I am wondering if this is the more straightforward option than what I've been suggesting so far.  I understand the `sitemap.xml` is the best way to broadcast to search engines what your content is, and so maybe this is a better approach if the goal is to communicate to search engines what is worth looking at?  Not too sure of the implications on link crawling etc though.

**Reply by slorber** [reply]
Hey

As far as I know the sitemap.xml is not like a RSS feed, and unless you submit it to Google manually or through an API there's no std way to tell Google with page meta what's the sitemap of your site. 

I'm not even sure it's that useful, apart if you have unlinked pages that the crawler can't find naturally, or want to speed up the indexing of some new pages (for which I think you can submit URLs as JSON which is a better solution, see https://developers.google.com/search/apis/indexing-api/v3/quickstart#sitemaps).

Afaik excluding content from the sitemap is not a signal for Google to not index those pages. You also have to ensure these pages were not previously discovered, and can't be found be crawling links. Adding a noindex with swizzle is probably a better solution and automatically filters the URLs from the sitemap.

**Reply by johnnyreilly** [reply]
I didn't know about indexing-api - thanks for sharing!  I explicitly removed my existing sitemap and resupplied afterwards to let Google know that things had significantly changed.

Part of the motivation was doing a search in Google and having a `/tags/` page come up with the post I was after rather than the actual blog post.

My hope is that over time Google will stop favouring `/tags/` and `/pages/` pages now I've delisted them.

**Reply by slorber** [reply]
I'm not 100% sure removing sitemap + submitting new one is a signal for google to stop indexing those pages. They are in the Google crawler DB now, so Google doesn't need a sitemap anymore to discover them, and will just schedule new crawls and keep them in the index until they go down? 🤔 but they will eventually have a lower ranking due to removed internal links

**Reply by johnnyreilly** [reply]
Yeah - I've another trick up my sleeve. I'm going to add a `noindex` to all `/page/` and `/tag/` pages. That should give Google sufficient signal to stop listing them

https://developers.google.com/search/docs/crawling-indexing/block-indexing

**Reply by johnnyreilly** [reply]
you can see what I'm getting up to here: https://github.com/johnnyreilly/blog.johnnyreilly.com/pull/350
